### ISSUE TITLE
Missing Hex() and DateDiffSecond() in 5.0.0-alpha3-ci

### ISSUE THREAD

**Issue by mc0re** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#1355]
### The issue
After switching to Pomelo 5.0.0 and EFCore 5.0.4, it looks like the `Hex()` and `DateDiffSecond()` mappings for in-memory database gone missing.

Message: 
> System.InvalidOperationException: An exception was thrown while attempting to evaluate the LINQ query parameter expression 'EF.Functions.Hex(value(CourseRepository+<>c__DisplayClass13_0).hash)'. See the inner exception for more information. ---> System.InvalidOperationException: The 'Hex' method is not supported because the query has switched to client-evaluation. This usually happens when the arguments to the method cannot be translated to server. Rewrite the query to avoid client evaluation of arguments so that method can be translated to server.

Stack trace:
```
Result StackTrace:	
at Microsoft.EntityFrameworkCore.MySqlDbFunctionsExtensions.Hex[T](DbFunctions _, T value)
   at lambda_method4052(Closure )
   at Microsoft.EntityFrameworkCore.Query.Internal.ParameterExtractingExpressionVisitor.GetValue(Expression expression, String& parameterName)
--- End of inner exception stack trace ---
    at Microsoft.EntityFrameworkCore.Query.Internal.ParameterExtractingExpressionVisitor.GetValue(Expression expression, String& parameterName)
   at Microsoft.EntityFrameworkCore.Query.Internal.ParameterExtractingExpressionVisitor.Evaluate(Expression expression, Boolean generateParameter)
   at Microsoft.EntityFrameworkCore.Query.Internal.ParameterExtractingExpressionVisitor.Visit(Expression expression)
   at System.Dynamic.Utils.ExpressionVisitorUtils.VisitArguments(ExpressionVisitor visitor, IArgumentProvider nodes)
   at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
   at Microsoft.EntityFrameworkCore.Query.Internal.ParameterExtractingExpressionVisitor.Visit(Expression expression)
   at System.Linq.Expressions.ExpressionVisitor.VisitLambda[T](Expression`1 node)
   at Microsoft.EntityFrameworkCore.Query.Internal.ParameterExtractingExpressionVisitor.Visit(Expression expression)
   at System.Linq.Expressions.ExpressionVisitor.VisitUnary(UnaryExpression node)
   at Microsoft.EntityFrameworkCore.Query.Internal.ParameterExtractingExpressionVisitor.Visit(Expression expression)
   at System.Dynamic.Utils.ExpressionVisitorUtils.VisitArguments(ExpressionVisitor visitor, IArgumentProvider nodes)
   at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
   at Microsoft.EntityFrameworkCore.Query.Internal.ParameterExtractingExpressionVisitor.Visit(Expression expression)
   at System.Dynamic.Utils.ExpressionVisitorUtils.VisitArguments(ExpressionVisitor visitor, IArgumentProvider nodes)
   at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
   at Microsoft.EntityFrameworkCore.Query.Internal.ParameterExtractingExpressionVisitor.Visit(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.Internal.ParameterExtractingExpressionVisitor.ExtractParameters(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteAsync[TResult](Expression query, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.ExecuteAsync[TResult](Expression expression, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ExecuteAsync[TSource,TResult](MethodInfo operatorMethodInfo, IQueryable`1 source, Expression expression, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ExecuteAsync[TSource,TResult](MethodInfo operatorMethodInfo, IQueryable`1 source, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.FirstOrDefaultAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
```

### Further technical details

MySQL version: In-memory
Operating system: Windows
Pomelo.EntityFrameworkCore.MySql version: 5.0.0-alpha.3.ci.20210223004403
Microsoft.AspNetCore.App version: 5.0.4

**Comment by lauxjpn** [comment]
That is expected. See [Provider-specific EF.Functions methods throw for InMemory provider](https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-5.0/breaking-changes#provider-specific-effunctions-methods-throw-for-inmemory-provider) in the EF Core docs for further information.

**Comment by mc0re** [comment]
Have they bloody gone nuts?!? How am I supposed to unit-test the functionality now? Spawn my own little _MySQL_ database for every test case???
Damn. Do you have any suggestions how one can mitigate this irresponsible change?

**Comment by lauxjpn** [comment]
> Spawn my own little MySQL database for every test case???

That's basically it. Depending on your tests, it might be enough to do it on a test class level tough (and use a test fixture that sets-up the database). It is what we do in our unit tests as well. I think it is the clean way to go forward, and should not be too hard as long as you have a local MySQL instance available where you run your tests, if you are already using the in-memory provider. But it will take a couple of hours.

> Do you have any suggestions how one can mitigate this irresponsible change?

Assuming you don't want to maintain your own version of Pomelo, that does implement the client-side behavior again, you could also use [Harmony](https://github.com/pardeike/Harmony) to hook into our method and redirect calls to them to your own class (which would just be the old class we were using before that implements the methods).

**Comment by mc0re** [comment]
@lauxjpn That's going to be awfully slow compared to the in-memory database :-(   And I'll need to install a MySQL instance - which is probably not an option in _Azure Devops_ / _GitHub_ / _GitLab_ unless I connect to the real database...

I also found this, don't know yet if I can apply it: https://github.com/matttamatic/MockQueryable/tree/InMemEFLike

Where can I find the old implementation?

**Comment by lauxjpn** [comment]
> That's going to be awfully slow compared to the in-memory database

Of the approaches I suggested, only the one with the actual MySQL store will be slower. The others will be as fast as fast as before, because they are still using the in-memory provider.

The ~18K test that we run against MySQL, take around 3-4 min. locally and about 7 min. on Azure DevOps, but they explicitly contain long-running (minutes) tests. So it could perform way better than you might expect. As always, you should test it first to find out where you stand.

> And I'll need to install a MySQL instance

That is perfectly possible both locally and on all common CI providers (take a look at our yaml script for some inspiration).

> Where can I find the old implementation?

Take a look at the the [history of the MySqlDbFunctionsExtensions.cs](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/commits/29dcac9398249c4bb3510423db294d4e927f966f/src/EFCore.MySql/Extensions/MySqlDbFunctionsExtensions.cs) file.

**Comment by lauxjpn** [comment]
Here is a quick sample program demonstrating how to simply continue to use the in-memory provider:

```c#
using System;
using System.Diagnostics;
using System.Linq;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;

namespace IssueConsoleTemplate
{
    public class IceCream
    {
        public int IceCreamId { get; set; }
        public string Name { get; set; }
        public DateTime ExpirationDate { get; set; }
    }

    public static class ClientSideEnabledMySqlDbFunctionsExtensions
    {
        public static int DateDiffYear(
            this DbFunctions _,
            DateTime startDate,
            DateTime endDate)
            => (int) ((endDate - startDate).TotalDays / 365);
        
        // ...
    }
    
    public class Context : DbContext
    {
        public virtual DbSet<IceCream> IceCreams { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            if (!optionsBuilder.IsConfigured)
            {
                optionsBuilder.UseInMemoryDatabase("Issue1355")
                    .UseLoggerFactory(
                        LoggerFactory.Create(
                            configure => configure
                                .AddConsole()
                                .AddFilter(level => level >= LogLevel.Information)))
                    .EnableSensitiveDataLogging()
                    .EnableDetailedErrors();
            }
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<IceCream>(
                entity =>
                {
                    entity.HasData(
                        new IceCream {IceCreamId = 1, Name = "Vanilla", ExpirationDate = new DateTime(2021, 12, 31)},
                        new IceCream {IceCreamId = 2, Name = "Chocolate", ExpirationDate = new DateTime(2023, 12, 31)});
                });
        }
    }

    internal static class Program
    {
        private static void Main()
        {
            using var context = new Context();

            context.Database.EnsureDeleted();
            context.Database.EnsureCreated();

            var today = new DateTime(2021, 4, 3);

            // Uses ClientSideEnabledMySqlDbFunctionsExtensions.DateDiffYear():
            var expiringWithin1Year = context.IceCreams
                .Where(i => EF.Functions.DateDiffYear(today, i.ExpirationDate) < 1)
                .ToList();
            
            Trace.Assert(expiringWithin1Year.Count == 1);
            Trace.Assert(expiringWithin1Year[0].IceCreamId == 1);
        }
    }
}
```

**Comment by lauxjpn** [comment]
And here is an example, how to rewrite an expression tree to use another extension method class then ours.
It should work in for all scenarios:

```c#
using System;
using System.Diagnostics;
using System.Diagnostics.CodeAnalysis;
using System.Linq;
using System.Linq.Expressions;
using System.Reflection;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.InMemory.Query.Internal;
using Microsoft.EntityFrameworkCore.Query;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;

namespace IssueConsoleTemplate
{
    public class IceCream
    {
        public int IceCreamId { get; set; }
        public string Name { get; set; }
        public DateTime ExpirationDate { get; set; }
    }
    
    public static class ClientSideEnabledMySqlDbFunctionsExtensions
    {
        public static int DateDiffYear(
            this DbFunctions _,
            DateTime startDate,
            DateTime endDate)
            => endDate.Year - startDate.Year;

        // ...
    }

    public class CustomInMemoryQueryableMethodTranslatingExpressionVisitorFactory : IQueryableMethodTranslatingExpressionVisitorFactory
    {
        private readonly QueryableMethodTranslatingExpressionVisitorDependencies _dependencies;
        
        public CustomInMemoryQueryableMethodTranslatingExpressionVisitorFactory(
            [NotNull] QueryableMethodTranslatingExpressionVisitorDependencies dependencies)
        {
            _dependencies = dependencies;
        }
        
        public virtual QueryableMethodTranslatingExpressionVisitor Create(QueryCompilationContext queryCompilationContext)
        {
            if (queryCompilationContext == null)
                throw new ArgumentNullException(nameof(queryCompilationContext));

            return new CustomInMemoryQueryableMethodTranslatingExpressionVisitor(_dependencies, queryCompilationContext);
        }
    }

    public class CustomInMemoryQueryableMethodTranslatingExpressionVisitor : InMemoryQueryableMethodTranslatingExpressionVisitor
    {
        public CustomInMemoryQueryableMethodTranslatingExpressionVisitor(
            [NotNull] QueryableMethodTranslatingExpressionVisitorDependencies dependencies,
            [NotNull] QueryCompilationContext queryCompilationContext)
            : base(dependencies, queryCompilationContext)
        {
        }

        public override Expression Visit(Expression expression)
        {
            expression = new RewritingExpressionVisitor().Visit(expression);
            return base.Visit(expression);
        }
    }

    public class RewritingExpressionVisitor : ExpressionVisitor
    {
        protected override Expression VisitMethodCall(MethodCallExpression expression)
        {
            if (expression.Method.DeclaringType == typeof(MySqlDbFunctionsExtensions))
            {
                var source = expression.Method;
                var runtimeMethods = typeof(ClientSideEnabledMySqlDbFunctionsExtensions)
                    .GetRuntimeMethods();
                var target = runtimeMethods
                    .FirstOrDefault(t => t.Name == source.Name &&
                                         t.ReturnParameter?.ParameterType == source.ReturnParameter?.ParameterType &&
                                         t.GetParameters().Select(p => p.ParameterType).SequenceEqual(source.GetParameters().Select(p => p.ParameterType)) &&
                                         t.GetGenericArguments() == source.GetGenericArguments());
                
                if (target != null)
                {
                    expression = Expression.Call(target, expression.Arguments);
                }
            }
            
            return base.VisitMethodCall(expression);
        }
    }
    
    public class Context : DbContext
    {
        public virtual DbSet<IceCream> IceCreams { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            if (!optionsBuilder.IsConfigured)
            {
                var serviceProvider = new ServiceCollection()
                    .AddEntityFrameworkInMemoryDatabase()
                    .AddLogging(
                        configure => configure
                            .AddConsole()
                            .AddFilter(level => level >= LogLevel.Information))
                    .AddSingleton<IQueryableMethodTranslatingExpressionVisitorFactory, CustomInMemoryQueryableMethodTranslatingExpressionVisitorFactory>()
                    .BuildServiceProvider();
                
                optionsBuilder.UseInMemoryDatabase("Issue1355_01")
                    .UseInternalServiceProvider(serviceProvider)
                    .EnableSensitiveDataLogging()
                    .EnableDetailedErrors();
            }
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<IceCream>(
                entity =>
                {
                    entity.HasData(
                        new IceCream {IceCreamId = 1, Name = "Vanilla", ExpirationDate = new DateTime(2021, 12, 31)},
                        new IceCream {IceCreamId = 2, Name = "Chocolate", ExpirationDate = new DateTime(2023, 12, 31)});
                });
        }
    }

    internal static class Program
    {
        private static void Main()
        {
            using var context = new Context();

            context.Database.EnsureDeleted();
            context.Database.EnsureCreated();

            var today = new DateTime(2021, 4, 3);

            // Forcing usage of MySqlDbFunctionsExtensions by explicitly calling MySqlDbFunctionsExtensions.DateDiffYear()
            // here.
            // Not necessary in the real world.
            // (Only needed in this sample code, because we want to demonstrate, how to rewrite an expression, even
            // though an EF.Functions.DateDiffYear() call would be perfectly fine here, because we defined the extension
            // method above (which has precedence over the one from MySqlDbFunctionsExtensions (Pomelo)).
            var expiringWithin1Year = context.IceCreams
                .Where(i => MySqlDbFunctionsExtensions.DateDiffYear(EF.Functions, today, i.ExpirationDate) < 1)
                .ToList();
            
            Trace.Assert(expiringWithin1Year.Count == 1);
            Trace.Assert(expiringWithin1Year[0].IceCreamId == 1);
        }
    }
}
```

**Comment by mc0re** [comment]
Thanks a lot, you've been most helpful.

**Comment by mc0re** [comment]
A couple of comments on the expression rewriting (which I chose).

1. In one case a `ParameterExtractingExpressionVisitor` was used instead of `RewritingExpressionVisitor`. I managed to modify the query (the call to `Hex` wasn't actually needed), but if anyone comes by this post, (s)he might need to add it.
2. The solution uses `MySqlStrings.ExpressionTypeMismatch`, which yields EF1001 "internal API" warning.
3. The same warning is produced by using `InMemoryQueryableMethodTranslatingExpressionVisitor`.

So all in all, this workaround over yet another Microsoft's disregard for developers worked for me, but it might be subject to expiry, when the giant decides to change something again.

**Comment by lauxjpn** [comment]
@mc0re Thanks for sharing!

> 1. In one case a `ParameterExtractingExpressionVisitor` was used instead of `RewritingExpressionVisitor`. I managed to modify the query (the call to Hex wasn't actually needed), but if anyone comes by this post, (s)he might need to add it.

Good point! This should only happen, if you (probably accidentally) call the `EF.Functions` method directly client-side (parameter related). Basically, EF Core realizes, that it is safe to call the `EF.Functions` client-side, because the expression in question does not rely on database related values. In that case, the `ParameterExtractingExpressionVisitor` will handle the `MethodCallExpression`.
Since it is not that simple to replace `ParameterExtractingExpressionVisitor` and the code is executed client-side anyway, I would suggest to change to expression in question in a way, that doesn't rely on the `EF.Functions` method at all, but just uses regular .NET Core method implementations instead (e.g. use a `String.Contains()` or a `Regex.IsMatch()` instead of `EF.Functions.Like()`).

> 2. The solution uses `MySqlStrings.ExpressionTypeMismatch`, which yields EF1001 "internal API" warning.
> 3. The same warning is produced by using InMemoryQueryableMethodTranslatingExpressionVisitor.
>
> [...] but it might be subject to expiry, when the giant decides to change something again.

Yes, seems to be unavoidable in this case, as class and interface are marked as _internal_. The warning can (and should be) suppressed with `#pragma warning disable EF1001`. It is unlikely, that there will be significant changes to the class in the future, so I think it is pretty safe to use this approach.

The main burden of continuing to use `EF.Functions` methods with the InMemory provider is, that you know have to maintain the client-side implementations yourself and provider your own implementations for added methods in the future, if you use them.

---

For others reading this issue, it should be mentioned, that this is not our official recommendation here. We also recommend to use a real database for unit testing (especially, if you are starting a new project).

This is just a working solution for anybody, that wants to continue to use the InMemory provider as before and is aware of possible divergences between the client-side implementation of an `EF.Functions` method and its server-side SQL counterpart. 

(Personally, I stopped breaking into peoples workplaces, to check whether their code complies with the latest recommendations, years ago.)

**Comment by mc0re** [comment]
> call the EF.Functions method directly client-side

I used something like
```
.Where(e => e.Refs.Select(r => Hex(r.Array)).Contains(Hex(reqArray))
```
Which I replaced with
```
.Where(e => e.Refs.Select(r => r.Array).Contains(reqArray)
```

> We ... recommend to use a real database
As long as this scenario requires installing an extra program on the machine, it will be inferior (user-friendly-wise) to the _InMemory_ provider, which only required a NuGet package. Or _SqLite_ in-memory solultion (which I couldn't get to work yet, but it's also an integral part of the respective NuGet). Of course, functionality check should be done on the real database. It's just not easy.
